src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -416,90 +343,96 @@ private BoundExpression RewriteStringConcatInExpressionLambda(SyntaxNode syntax,
 
-                        // Can the expression be optimized with a ToString call?
-                        // If so, we can synthesize a ToString call to avoid boxing.
-                        if (ConcatExprCanBeOptimizedWithToString(operand.Type))
-                        {
-                            var toString = UnsafeGetSpecialTypeMethod(syntax, SpecialMember.System_Object__ToString);
+            // If it's a string already, just return it
+            if (expr.Type.SpecialType == SpecialType.System_String)
+            {
+                return expr;
+            }
 
-                            var type = (NamedTypeSymbol)operand.Type;
-                            var toStringMembers = type.GetMembers(toString.Name);
-                            foreach (var member in toStringMembers)
-                            {
-                                var toStringMethod = member as MethodSymbol;
-                                if (toStringMethod.GetLeastOverriddenMethod(type) == (object)toString)
-                                {
-                                    return BoundCall.Synthesized(syntax, operand, toStringMethod);
-                                }
-                            }
-                        }
+            // Evaluate toString at the last possible moment, to avoid spurious diagnostics if it's missing.
+            // All code paths below here use it.
+            var toString = UnsafeGetSpecialTypeMethod(syntax, SpecialMember.System_Object__ToString);
+
+            // If it's a special value type, we know that it has its own ToString method. Assume that this won't
+            // be removed, and emit a direct call rather than a constrained virtual call. 
+            // This keeps in the spirit of #7079, but expands the range of types to all special value types.
+            if (expr.Type.IsValueType && expr.Type.SpecialType != SpecialType.None)
+            {
+                var type = (NamedTypeSymbol)expr.Type;
+                var toStringMembers = type.GetMembers(toString.Name);
+                foreach (var member in toStringMembers)
+                {
+                    var toStringMethod = (MethodSymbol)member;
+                    if (toStringMethod.GetLeastOverriddenMethod(type) == (object)toString)
+                    {
+                        return BoundCall.Synthesized(expr.Syntax, expr, toStringMethod);
                     }
                 }
             }
 
-            // Optimization not possible; just return the original expression.
-            return expr;
-        }
+            // If it's a value type (or unconstrained generic), and it's not constant and not readonly,
+            // then we need a copy. This is to mimic the old behaviour, where ToString was called on a box
+            // of the value type, and so any side-effects of ToString weren't made to the original.
+            if (!expr.Type.IsReferenceType && !expr.Type.IsReadOnly && expr.ConstantValue == null)
```

Comment:
Suggest checking the `IsEffectivelyReadOnly` of this type's `ToString` method instead of checking the type's `IsReadOnly`.

Reply:
I did not know that was a thing, will do! Presumably if the type doesn't have a `ToString` method we fall back to the type's readonlyness?

Reply:
Here's my understanding on when we need to pass by copy to prevent the caller from observing mutations:

1. We are calling `System.Object.ToString()` on a struct. No copy is needed.
2. We are calling a non-readonly ToString defined on the struct. We need to copy.
3. We are calling a readonly ToString defined on the struct. No copy is needed.
4. We are calling a ToString defined on a readonly struct. No copy is needed. The method symbol will have `IsEffectivelyReadOnly == true`.

https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACBmAAlgQFcBjGAgYWjgEYBYAKAG9mCOCc6A2LgFgIAVOLAAUaAHbVadAmToBKdpzZNOG+XQB0QgPYBlGAikBzMYoDcKjgF9m9ps3xFj5SjShwATMzUaXPQA3RBMAEzguOgAGYUNjMwsCAF4APgIAIk8fDOt1ThsovhxBEXEpGS9veW9lfI5/TQ4yb114k0lzK0LHR2dCYndKuDw/QpcEOAgwvUkMAE8CYNC0CKjY/SMOrpT0rNo8XMLx3gFhURgJaWzCMjw6jUamu7atxO76gl7mfoJJ6dmC1cpAow34Y0+gRCCHCkW4G3a712mRuR0+J2KpQuVzB8n4D1UhQ0ZH4rwSnQseQ03yYQA==

Reply:
I'm worried about case 1 because of versioning. The user could add a ToString method later which has side-effects. That's why the code currently emits a copy.

Reply:
IMO the versioning concern is applicable even if the struct is readonly-- it's not a binary break to remove 'readonly', so someone could compile against a 'readonly' struct and then load a non-'readonly' one from a different version of the assembly at runtime. I think we might be a little more relaxed with this because it's innately so easy to break compatibility (binary or behavioral) when modifying a struct in another assembly.

At the low level, this comes down to the BoundCall to System.Object.ToString() requiring only a readable address instead of a writable one for the receiver. I'm not 100% sure why it is that way today.

perhaps @agocke @vsadov may know?

Reply:
Right. My assumption was that removing "readonly" would be a binary break.

I think this is an unlikely case either way: people concatenate structs with strings because they're interested in the return value of the struct's ToString method. If there's no ToString method, I can't see why someone would do the concatenation. So I'd be very surprised if this ever mattered in practice.

If you're sure, I'll make the change, no problem. 

I'm a little worried that the compiler's noticed that the struct and the method using it are in the same assembly so one couldn't change without the other being recompiled. But if it was that smart it would have used call instead of a constrained virtual call I suppose...

Reply:
> At the low level, this comes down to the BoundCall to System.Object.ToString() requiring only a readable address instead of a writable one for the receiver. I'm not 100% sure why it is that way today.

My understanding was that the semantics caused Concat to box any struct argument. In that case, the writeability doesn't matter